scripts/generateRelatedAssetIndex/generateRelatedAssetIndex.ts

```diff
@@ -0,0 +1,127 @@
+    const url = '/fungibles'
+    const payload = { ...options, ...filter, url }
+    const { data: res } = await axios.request(payload)
+    const validationResult = zerionFungiblesSchema.parse(res)
+    const firstEntry = validationResult.data[0]
+
+    if (firstEntry === undefined) return
+
+    const implementations = firstEntry.attributes.implementations
+    const primaryImplementationId = firstEntry.id
+
+    const primaryImplementation = implementations?.find(
+      implementation => implementation.address === primaryImplementationId,
+    )
+
+    const relatedAssetKey = primaryImplementation
+      ? zerionImplementationToMaybeAssetId(primaryImplementation)
+      : undefined
+
+    const relatedAssetIds = implementations?.map(zerionImplementationToMaybeAssetId).filter(isSome)
+
+    // skip empty result as there is no value adding an empty lookup
+    // skip singleton result as there is no value having a lookup to only itself
+    if (!relatedAssetKey || !relatedAssetIds || relatedAssetIds.length <= 1) {
+      return
+    }
+
+    return { relatedAssetIds, relatedAssetKey }
+  } catch (e) {
+    console.error(e)
+    throw e
+  }
+}
+
+const processRelatedAssetIds = async (
+  assetId: AssetId,
+  assetData: AssetsById,
+  relatedAssetIndex: Record<AssetId, AssetId[]>,
+): Promise<void> => {
+  // don't fetch if we've already got the data from a previous request
+  const existingRelatedAssetKey = assetData[assetId].relatedAssetKey
+  if (existingRelatedAssetKey) {
+    return
+  }
+
+  const relatedAssetsResult = await getRelatedAssetIds(assetId)
+
+  if (!relatedAssetsResult) return
+
+  const { relatedAssetIds, relatedAssetKey } = relatedAssetsResult
+
+  for (const assetId of relatedAssetIds) {
+    assetData[assetId].relatedAssetKey = relatedAssetKey
+  }
+  relatedAssetIndex[relatedAssetKey] = relatedAssetIds
+}
+
+const generateRelatedAssetIndex = async () => {
+  console.log('generateRelatedAssetIndex() starting')
+
+  const generatedAssetsPath = path.join(
+    __dirname,
+    '../../src/lib/asset-service/service/generatedAssetData.json',
+  )
+  const relatedAssetIndexPath = path.join(
+    __dirname,
+    '../../src/lib/asset-service/service/relatedAssetIndex.json',
+  )
+
+  const generatedAssetData: AssetsById = require(generatedAssetsPath)
+  const relatedAssetIndex: Record<AssetId, AssetId[]> = {}
+  const assetDataWithRelatedAssetKeys: AssetsById = { ...generatedAssetData }
+
+  for (const assetId of Object.keys(generatedAssetData).slice(0, 50)) {
+    await processRelatedAssetIds(assetId, assetDataWithRelatedAssetKeys, relatedAssetIndex)
+  }
+
+  fs.writeFileSync(
+    generatedAssetsPath,
+    // beautify the file for github diff.
+    JSON.stringify(assetDataWithRelatedAssetKeys, null, 2),
+  )
+
+  fs.writeFileSync(
+    relatedAssetIndexPath,
+    // beautify the file for github diff.
+    JSON.stringify(relatedAssetIndex, null, 2),
+  )
+}
+
+generateRelatedAssetIndex()
```

Comment:
noting here so i dont forget - we will need to chain this with the existing asset generation so we never have a case where the asset list gets updated without this (since this modifies the generated asset list)